### ISSUE TITLE
Question: Parallel writes

### ISSUE THREAD

**Issue by DonIvanCorleone** [cwida/duckdb#867]
Hello,

first of all: Awesome project! :heart: it 
I was pretty eager to play around with the possibilities and came across one question: Would it be possible to have parallel writes using duckdb?

A short tour through your documentation did not mention anything against it (at least I did not see it), so I just gave it a try. 

``` import pyarrow.plasma as plasma
import duckdb
import os
import concurrent.futures

def f(id):
    print("Start processing with nr: {}".format(id))
    connection = duckdb.connect(database='/tmp/mem', read_only=False)
    print("Inserting ....")
    connection.execute("INSERT INTO test SELECT * FROM parquet_scan('/path/to/MyFile.parquet');")
    connection.commit()
  

if __name__ == '__main__':

    try:
        os.remove('/tmp/mem')
        os.remove('/tmp/mem.wal')
    except:
        print("Nothing to remove")

    con = duckdb.connect(database='/tmp/mem', read_only=False)
    
    con.execute("CREATE TABLE test AS SELECT * FROM parquet_scan('/path/to/MyFile.parquet');")
    con.execute("INSERT INTO test SELECT * FROM parquet_scan('/path/to/MyFile.parquet');")
    con.commit()
   

    with concurrent.futures.ProcessPoolExecutor(max_workers=10) as executor:
        futures = {executor.submit(f, val): val for val in [1,2,3]}

        for future in concurrent.futures.as_completed(futures):
            try:
                data = future.result()
            except Exception as exc:
                print('generated an exception: %s' % (exc))
```

This quick test to write in parallel processes resulted in the following error message:

**IO: Could not set lock on file "/tmp/mem.wal": Resource temporarily unavailable**

To be honest I somehow expected this behaviour, but I am still curious to ask you guys the following: 

1. Is this behaviour expected and intended?
2. Are there any plans to make parallel writes from different processes  possible?
3. Do you have any suggestions to circumvent this problem? - besides doing everything sequentially of course :)

Many thanks in advance
Looking forward to your answer

Cheers

**Comment by Mytherin** [comment]
Parallel writes from concurrent processes are unfortunately not planned right now. They impose a lot of restrictions from an efficiency standpoint, because basically nothing can be cached in memory. Everything needs to be written to the file immediately so other processes can correctly see the changes. SQLite does support this, but implements it in the manner described above (every individual write locks the entire file - applies the change - flushes it to disk - then unlocks the file).

While it would be a nice to have feature, the implications from a performance and system design perspective are big, and therefore we are not planning on making this possible in the future either.

A work-around would either be to use DuckDB as a "regular" client-server database, and wrap DuckDB with a client-server protocol. The server can then accept clients and perform writes as a single process. 

Another alternative would be to manually implement this flush-after-every-write behavior by "opening a connection to a database file - performing the writes - closing the connection again". The system is not particularly optimized for this scenario, however.

**Comment by DonIvanCorleone** [comment]
Thanks a lot for your answer! I see your points and agree.
 
Thanks for your suggestions/ideas, I think I will check how far I will come with a manual implementation... Wish me luck :)

**Comment by Mytherin** [comment]
Let us know how it goes :) We have a very simple REST-style client/server already (tools/rest), but we have thought about adding a Postgres-protocol compatible client-server at some point in the future. I'm not sure if that would solve your problem or not.

**Comment by qiuwei** [comment]
Is it possible to lock on the table level? So that concurrent writes to different tables are allowed.
